### ISSUE TITLE
Neural Network Feature

### ISSUE THREAD

**Issue by thomasnield** [optimatika/ojAlgo#101]
Just a thought. I've been messing around with ojAlgo to back a neutral network. Considering there is already an MIP solver, a neural network API would also be cool and this library seems to have all the pieces needed for it.

Here's what I've been doing so far.

https://github.com/thomasnield/kotlin_simple_neural_network/blob/master/src/main/kotlin/NeuralNetwork.kt

**Comment by apete** [comment]
I have thought about it. More precisely I thought about building stuff to enable reconstructing/using already trained/tuned networks. I assume the various existing neural network libraries come with some sort mechanism to store (and later recreate) network models.

ojAlgo could contain the basics for how to construct and use a neural network, and ojAlgo-extensions could contain extensions to various 3:d party libraries that would enable using their trained models in pure Java. Something like that...

The only thing missing is the time to do it. ;-)

**Comment by thomasnield** [comment]
I'll contribute if I have something obviously useful. I'll figure out the scope as I go, as we obviously aren't going re-engineer TensorFlow. But a nice, simple, friendly neural network API would be great.

**Comment by apete** [comment]
I created a feature branch to play around with some code.

https://github.com/optimatika/ojAlgo/tree/feature/ArtificialNeuralNetwork

The optimisation algorithms already in ojAlgo are direct applicable to ANN training, but think it won't be too much work to add what's needed.

**Comment by thomasnield** [comment]
@apete fun! Alright I'll play with this this week.

**Comment by thomasnield** [comment]
I kind of wondered about that optimization part. It almost feels too easy to use an MIP solver but it seems correct in spirit.

Im wondering if nonlinear activation functions could be a problem using an MIP solver, but the solver would only be used to calculate weights for training. Activation functions like Sigmoid may not have to be passed to the MIP because the node values may already be calculated.

**Comment by apete** [comment]
Now I think it's possible to play around with it...

**Comment by thomasnield** [comment]
@apete you're giving it a shot?

**Comment by thomasnield** [comment]
Excellent, I have an immediate need for a simple ANN solution so I'll give this a spin later. I'm not satisfied with my toy example which already is using ojAlgo. 

https://github.com/thomasnield/kotlin_simple_neural_network/

**Comment by thomasnield** [comment]
This works fantastically. I used it in my demo app to predict dark/light fonts for a given background color. Quick question? Is this using stochastic gradient descent? 

Thanks again for this fun feature. 

https://github.com/thomasnield/kotlin_simple_neural_network

**Comment by apete** [comment]
It's gradient descent in its simplest form - no batching, BFGS or anything. I believe that's called stochastic gradient descent. I've focused on making each training iteration as "cheap" as possible. Training seems to work best with a lower learning rate and iterating over the training set several times.

**Comment by thomasnield** [comment]
Okay good to know. I did notice the immutable nature of the implementation. Personally I like that a lot. But I'm sure other folks are going to ask for support in adding more training data over time.

I really think this could be a killer feature for ojAlgo. I tried DL4J but it was too heavyweight and unusable for toy examples. This however is easy to use. It also conveniently contains the linear algebra implementation too.
